daemon/src/lib.rs

```diff
@@ -200,17 +202,18 @@ where
 
         let (endpoint_addr, endpoint_context) = Context::new(None);
 
-        let ping_address = xtra_libp2p_ping::Actor::new(endpoint_addr.clone(), PING_INTERVAL)
+        ping::Actor::new(endpoint_addr.clone(), PING_INTERVAL)
```

Comment:
It's a nifty way of getting around of congestion in ping, but I must admit am not in favour of this commit  - it just hides the underlying problem (congestion on processing external messages and trying to handle internal ones).

I feel like we need a more general solution, as this cannot be used in any protocols that require more interaction (e.g. rollover). 

the ping protocol problems we've seen is only a manifestation of problems that we see in other places, and should not be treated specially. We're looking at it as it's more isolated and easier to reproduce these problems - we can't roll such an approach to cfd actors etc.

We'll likely need a new way of scheduling internal messages in an async way, or a 2-tier queue (internal messages taking priority over external, or something else entirely!).

 out of curiosity: have you tried disabling prometheus metrics for the tests? has that changed anything?

Reply:
> It's a nifty way of getting around of congestion in ping, but I must admit am not in favour of this commit - it just hides the underlying problem (congestion on processing external messages and trying to handle internal ones).

I see no benefit in having one actor for this protocol. The split shows that the `ping` and `pong` actors share no code. They are doing different things. They don't need to share state (the `pong` actor is stateless). It makes no sense to block the processing of pings on processing messages that access internal state that we don't care about!

I think the original design was flawed because it unnecessarily entangled different concerns into a single actor. It may be attractive to put everything in the same actor, but we have to have a strong reason to do so. After all, an actor is just a mechanism to perform actions concurrently based on state.

> I feel like we need a more general solution, as this cannot be used in any protocols that require more interaction (e.g. rollover).
>
> the ping protocol problems we've seen is only a manifestation of problems that we see in other places, and should not be treated specially. We're looking at it as it's more isolated and easier to reproduce these problems - we can't roll such an approach to cfd actors etc.

My learning from this patch is that we should try to build smaller actors. I suspect that we can see performance improvements by applying this idea to other parts of the codebase.

> We'll likely need a new way of scheduling internal messages in an async way, or a 2-tier queue (internal messages taking priority over external, or something else entirely!).

I am absolutely for introducing extensions to `xtra`, either by contributing or building on top! I think there is potential to make performance improvements that way too.

> out of curiosity: have you tried disabling prometheus metrics for the tests? has that changed anything?

I did, but I saw no notable improvements :(

Reply:
> I think the original design was flawed because it unnecessarily entangled different concerns into a single actor. It may be attractive to put everything in the same actor, but we have to have a strong reason to do so. After all, an actor is just a mechanism to perform actions concurrently based on state.

an actor is an abstraction, and in this case I think the original design nailed the level of abstraction - abstracting one protocol per actor. IMHO this should not cause trouble if we didn't have some underlying problems. 

In recent work we started to avoid internal dispatch to actually achieve some concurrency - otherwise, if all the futures need to fan back in 1 by 1 to complete an internal dispatch, we love all the benefits! 

thanks for checking disabling the metrics, at least that tells us it's not the issue :-)

Reply:
> an actor is an abstraction, and in this case I think the original design nailed the level of abstraction - abstracting one protocol per actor. IMHO this should not cause trouble if we didn't have some underlying problems.

Sorry but I think this is wrong.

What is the benefit of putting the dialer and listener protocols in the same actor (other than code organisation, which is subjective), at the cost of fundamentally limiting our ability to perform the listener protocol in a timely manner?

By combining both protocols we are using one message queue when we need two.

Even the design of `xtra_libp2p` guides us towards only registering the listener protocol actor when constructing `Endpoint`: the dialer protocol works without passing its `xtra::Address` to the constructor.